files modified to create command line tasks cosi_bindata

ldigesu opened the pull request:
Files modified to create command line task cosi_bindata.
Added functions to automate creation of the input ymal file and binning.
Modified setup.py to allow call froma command line

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
files modified to create command line tasks cosi_bindata

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ldigesu committed:
Correct a typo

ckarwin commented:
@ldigesu Is this PR ready to be reviewed?

ldigesu commented:
> @ldigesu Is this PR ready to be reviewed?

@ckarwin Hello Chris, yes, to me this is ready for review. tks

ckarwin commented:
Ok, I will take a look.

ldigesu committed:
Merge branch 'cosipy_pipeline' of https://github.com/cositools/cosipy into cosipy_pipeline

ckarwin commented:
Hi @ldigesu I started reviewing your PR. Right away I ran into a problem.

First, I tested `cosi-threemlfit --config test_pipeline.yaml`, using the yaml file from `test_data`. This works fine. 

But then I tried `cosi-bindata --config test_pipeline.yaml`, and I get the following error:

```
Traceback (most recent call last):
  File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/lib/python3.10/site-packages/yayc/configurator.py", line 56, in __getitem__
    value = value[key]
KeyError: 'prepbdata'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/bin/cosi-bindata", line 7, in <module>
    sys.exit(cosi_bindata())
  File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/cosipy/cosipy/pipeline/task/task.py", line 70, in cosi_bindata
    config = Configurator(full_config[args.config_group])
  File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/lib/python3.10/site-packages/yayc/configurator.py", line 58, in __getitem__
    raise KeyError("Parameter '{}' doesn't exists".\
KeyError: "Parameter 'prepbdata' doesn't exists"
```

I'm running the same thing that's in the unit test, but for some reason I get this error. Am I doing something wrong? Do you get the same thing?

Another major comment on this PR is that I think you should make a simple example tutorial of how to use the apps. This can be part of the tutorials. It probably makes sense to make a new directory for apps. 

My final major comment, your code needs documentation, for each function. The doc string convention should follow Numpy/Scipy standards. If you want an example, you can see the doc strings here: https://github.com/cositools/cosipy/blob/develop/cosipy/data_io/BinnedData.py. And here is a guide for reference: https://numpy.org/doc/1.19/docs/howto_document.html.

ldigesu committed:
Added in task.py the option of adding a custom suffix to the names of the output files.

ldigesu committed:
Added in task.py the option of adding a custom suffix to the names of the output files.

ldigesu committed:
Added in task.py the option of adding a custom suffix to the names of the output files.

ldigesu committed:
Added in task.py the option of adding a custom suffix to the names of the output files.

ldigesu commented:
> Hi @ldigesu I started reviewing your PR. Right away I ran into a problem.
> 
> First, I tested `cosi-threemlfit --config test_pipeline.yaml`, using the yaml file from `test_data`. This works fine.
> 
> But then I tried `cosi-bindata --config test_pipeline.yaml`, and I get the following error:
> 
> ```
> Traceback (most recent call last):
>   File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/lib/python3.10/site-packages/yayc/configurator.py", line 56, in __getitem__
>     value = value[key]
> KeyError: 'prepbdata'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/bin/cosi-bindata", line 7, in <module>
>     sys.exit(cosi_bindata())
>   File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/cosipy/cosipy/pipeline/task/task.py", line 70, in cosi_bindata
>     config = Configurator(full_config[args.config_group])
>   File "/project/ckarwin/astro/chris/my_envs/cosipy_pr2/lib/python3.10/site-packages/yayc/configurator.py", line 58, in __getitem__
>     raise KeyError("Parameter '{}' doesn't exists".\
> KeyError: "Parameter 'prepbdata' doesn't exists"
> ```
> 
> I'm running the same thing that's in the unit test, but for some reason I get this error. Am I doing something wrong? Do you get the same thing?
> 
> Another major comment on this PR is that I think you should make a simple example tutorial of how to use the apps. This can be part of the tutorials. It probably makes sense to make a new directory for apps.
> 
> My final major comment, your code needs documentation, for each function. The doc string convention should follow Numpy/Scipy standards. If you want an example, you can see the doc strings here: https://github.com/cositools/cosipy/blob/develop/cosipy/data_io/BinnedData.py. And here is a guide for reference: https://numpy.org/doc/1.19/docs/howto_document.html.

ldigesu commented:
Hello @ckarwin, I have modified and committed the file task.py in a way that should fix the error you were getting (there was a naming mismatch). Let me know if it works for you, as it works fine for me now.  If yes, then I can address the other comments. The thing in the unit test should create two files named bin.yaml and binned_data_local.hdf5. I have also done something in PyCharm that apparently updated my branch, at least it looks so to me. tks and cheers, Laura

israelmcmc commented:
I'm closing and reopening this PR to see if the files changed are refreshed

ckarwin commented:
Hi @ldigesu, The cosi-bindata app is working for me now. Please let me know when you are ready for me to check the other updates.

ldigesu committed:
Merge remote-tracking branch 'upstream/cosipy_pipeline' into cosipy_pipeline

ldigesu committed:
Removed unnecessary build_spectrum function.
Modified tslice function to work with Time objects in input.
Removed unnecessary tslice_ori function and modified task.py accordingly.

ldigesu committed:
Removed unnecessary build_spectrum function.
Modified tslice function to work with Time objects in input.
Removed unnecessary tslice_ori function and modified task.py accordingly.

ldigesu commented:
Hello @ckarwin and @israelmcmc. I have now committed files including documentation for all the functions, and updated the branch using pycharm. About adding a tutorial demonstrating the usage of the apps as suggested by @ckarwin : what did you have in mind exactly?  I was thinking the following. I have a working example in my computer, which uses the apps to bin and cut in time a dc3 background file, a dc3 grb file, and then fit the grb in the whole time window, in the first ten seconds and then in the last twenty seconds. This is a 5 lines bash script (copied below), that could be turned in a tutorial notebook with explanations of each line. However, in a tutorial notebook we would need also to download all the files and prepare the data internally. Could something like this work or do you have other suggestions?


#
cosi-bindata --config prep_bk_fit_grbdc3.yaml --overwrite --suffix "galbk"
cosi-bindata --config prep_fit_grbdc3.yaml --overwrite --suffix "grbdc3"
#
cosi-threemlfit --config prep_fit_grbdc3.yaml --overwrite   --suffix "grbdc3"
#
cosi-threemlfit --config prep_fit_grbdc3.yaml --overwrite --tstart 1836496300.0 --tstop 1836496310.0   --suffix "grbdc3_first_ten_seconds"
cosi-threemlfit --config prep_fit_grbdc3.yaml --overwrite --override cuts:kwargs:tstart=1836496368.1 cuts:kwargs:tstop=1836496388.1 --suffix "grbdc3_last_twenty_seconds"

israelmcmc commented:
Thanks @ldigesu! 

I agree it would be nice to have usage examples for each app. These will likely not be Jupyter notebook, since the apps are meant to be called from the terminal. You can directly add these usage example as an `.rst` files ([here](https://github.com/cositools/cosipy/blob/develop/docs/api/polarization.rst)'s an example). I think this deserves a new folder in `docs` called `apps` or something like that. 

The example you described would be enough. You can indicate to the user to use `fetch_wasabi` to download the necessary files for the example.

ldigesu committed:
Removed unnecessary build_spectrum function.
Modified tslice function to work with Time objects in input.
Removed unnecessary tslice_ori function and modified task.py accordingly.

ldigesu committed:
Removed unnecessary build_spectrum function.
Modified tslice function to work with Time objects in input.
Removed unnecessary tslice_ori function and modified task.py accordingly.

ldigesu committed:
Added tutorial files and explanation.
Added documentations for functions.

ldigesu committed:
Added tutorial files and explanation.
Added documentations for functions.

ldigesu committed:
Added tutorial files and explanation.
Added documentations for functions.

ldigesu committed:
Added tutorial files and explanation.
Added documentations for functions.

ldigesu committed:
remove unecessary line from test file

ldigesu committed:
remove unecessary line from test file

ldigesu committed:
Merge remote-tracking branch 'upstream/cosipy_pipeline' into cosipy_pipeline

ldigesu commented:
Hello @israelmcmc and @ckarwin. I have now added the tutorial with an .rst explanation file in docs/tutorial/apps. I already added the documentation for all the functions. So I think you can proceed with the review. I also did a sync_fork in pycharm to update my branch. Let me know if you can proceed now. cheers,

ckarwin commented:
Thanks, @ldigesu. I plan to do my review in the next day or two.

ckarwin commented:
@israelmcmc, This PR is showing many commits and file changes that I think were already merged. Can you try resetting the PR again (i.e. close and reopen), to make it more clear which files have been changed?

israelmcmc commented:
I'll give it a try

israelmcmc commented:
@ckarwin I think it worked

ckarwin commented:
@israelmcmc thank you!

ckarwin commented on docs/tutorials/response/SpacecraftFile.ipynb:
```diff
@@ -86,39 +94,44 @@
    "source": [
     "ori_path = data_dir/\"20280301_3_month_with_orbital_info.ori\"\n",
-    "response_path = data_dir/\"SMEXv12.Continuum.HEALPixO3_10bins_log_flat.binnedimaging.imagingresponse.h5\""
+    "response_path = data_dir/\"SMEXv12.Continuum.HEALPixO3_10bins_log_flat.binnedimaging.imagingresponse.nonsparse_nside8.area.good_chunks_unzip.h5\""
```
@ldigesu Did you intentionally change this tutorial? I'm not sure why it needed to be changed for this PR.

ckarwin commented on docs/tutorials/response/DetectorResponse.ipynb:
```diff
@@ -14,10 +14,10 @@
    "metadata": {},
    "source": [
     "The detector response provides us with the the following information:\n",
-    "- The effective area at a given energy for a given direction. This allows us to convert from counts to physical quantities like flux\n",
```
@ldigesu Same here. Did you intentionally change this tutorial? It seems like a lot of the changes are just introducing new typos in the text.

ckarwin reviewed (commented):
I tested both python scripts and the shell script in tutorials/apps. They worked well. 

A couple suggestions for apps.rst: 

authomatize —> automate 

A practical example:  
The example scripts are very helpful. They contain multiple app calls, with subsequent calls relying on output from previous calls. It would be helpful to make it clear in the text that the example scripts must be ran as provided, since the inputs for subsequent app calls depend on previous calls. For example, I tried to copy just one of the calls and run it from the terminal, but I got a missing file error, since I didn’t first run the part above. 

I noticed changes to the DetectorResponse and SpaceCraftFile tutorials, and left specific comments there. Did you mean to change these? If so, I will run them quickly to verify they are ok b/c it seems like you changed some of the files that are being used. Otherwise, can you restore the tutorials to the original format?

ldigesu committed:
Addressed CK comment.

ldigesu committed:
Restored DetectorResponse.ipynb and SpacecraftFile.ipynb to version in develop. They were changed by mistake.

ldigesu commented:
@ckarwin tks for the comment. I committed a new version of the app.rst. I also restored the other two files to original format in develop. Indeed they were changed by accident.

ckarwin reviewed (approved):
Thanks @ldigesu. I think everything looks good now and it can be merged.
